api/utils/keys/yubikey.go

```diff
@@ -0,0 +1,407 @@
+	// This value may change between OS's, or with other system changes.
+	card string
+	// serialNumber is the yubikey's 8 digit serial number.
+	serialNumber uint32
+}
+
+func newYubikey(card string) (*yubikey, error) {
+	y := &yubikey{card: card}
+
+	yk, err := y.open()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer yk.Close()
+
+	y.serialNumber, err = yk.Serial()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return y, nil
+}
+
+// generatePrivateKey generates a new private key from the given PIV slot with the given PIV policies.
+func (y *yubikey) generatePrivateKey(slot piv.Slot, touchPolicy piv.TouchPolicy) (*YubikeyPrivateKey, error) {
+	yk, err := y.open()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer yk.Close()
+
+	opts := piv.Key{
+		Algorithm:   piv.AlgorithmEC256,
+		PINPolicy:   piv.PINPolicyNever,
+		TouchPolicy: touchPolicy,
+	}
+	pub, err := yk.GenerateKey(piv.DefaultManagementKey, slot, opts)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Create a self signed certificate and store it in the PIV slot so that other
+	// Teleport Clients know to reuse the stored key instead of genearting a new one.
+	priv, err := yk.PrivateKey(slot, pub, piv.KeyAuth{})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	cert, err := selfSignedTeleportClientCertificate(priv, pub)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Store an unsigned certificate to mark this slot as used by tsh.
+	if err = yk.SetCertificate(piv.DefaultManagementKey, slot, cert); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return newYubikeyPrivateKey(y, slot, pub)
+}
+
+// getPrivateKey gets an existing private key from the given PIV slot.
+func (y *yubikey) getPrivateKey(slot piv.Slot) (*YubikeyPrivateKey, error) {
+	yk, err := y.open()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer yk.Close()
+
+	// Check the slot's certificate to see if it contains a self signed Teleport Client cert.
+	cert, err := yk.Certificate(slot)
+	if err != nil || cert == nil {
+		return nil, trace.NotFound("Yubikey certificate slot is empty, expected a Teleport Client cert")
+	} else if len(cert.Subject.Organization) == 0 || cert.Subject.Organization[0] != certOrgName {
+		return nil, trace.NotFound("Yubikey certificate slot contained unknown certificate:\n%+v", cert)
+	}
+
+	// Attest the key to make sure it hasn't been imported.
+	slotCert, err := yk.Attest(slot)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	attestationCert, err := yk.AttestationCertificate()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if _, err = piv.Verify(attestationCert, slotCert); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Verify that the slot's certs have the same public key, otherwise the key
+	// may have been generated by a non-teleport client.
+	if !cert.PublicKey.(interface{ Equal(x crypto.PublicKey) bool }).Equal(slotCert.PublicKey) {
```

Comment:
Do we have a hard guarantee that the `PublicKey` for a `x509.Certificate` can only ever be a valid (the interface as described by the comment for) `crypto.PublicKey`?

Reply:
Actually no, because looking through the code path it could potentially be a `*dsa.PublicKey` which doesn't implement `Equal`. I'll add the check, thanks.